### PR TITLE
Proof of concept to replace the HTTP server with WKURLSchemeHandler

### PR DESCRIPTION
**Pull request by mickael-menu**
A first stab at https://github.com/readium/r2-navigator-swift/issues/148

There are a few issues to address.

#### Cancelled `WKURLSchemeTask`

The scheme handler is called on the main thread, so we **must** spawn a background task to read the resource and notify back the `WKURLSchemeTask`. However, if we received a cancellation notification for this task, any new call to `WKURLSchemeTask` crashes the app (low-level exception). But guarding the access to the `isCancelled` property with a serialized queue creates some random deadlocks.

#### Byte range requests

We don't get any byte range information, however the `WKURLSchemeTask` is cancelled before loading the full video content when first loading the page. We can also serve the video progressively so there's no need to load everything in memory at once, however I didn't find any way to start streaming from a particular offset. Need some more testing with large videos to see if random access would work fine.

According to [this bug report](https://bugs.webkit.org/show_bug.cgi?id=203302) on Mac we get byte range info but not on iOS. [This Twitter thread](https://twitter.com/Callionica/status/1255906555599425537) provides more context.

### CODE REVIEW

File: r2-navigator-swift/EPUB/EPUBSpreadView.swift [hunk 1]
```diff
@@ -64,16 +64,33 @@ class EPUBSpreadView: UIView, Loggable, PageView {
...
         self.spread = spread
         self.resourcesURL = resourcesURL
         self.readingProgression = readingProgression
         self.userSettings = userSettings
         self.editingActions = editingActions
         self.animatedLoad = animatedLoad
-        self.webView = WebView(editingActions: editingActions)
         self.contentInset = contentInset
+        if #available(iOS 11.0, *) {
+            let scheme = "readium-pub"
```
Comment (ettore) [inline r2-navigator-swift/EPUB/EPUBSpreadView.swift:86]: can this literal be a constant on the super class, since it's also used in `EPUBReflowableSpreadView`?

File: r2-navigator-swift/Toolkit/WebViewResourceHandler.swift [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        }
+        guard url.scheme == scheme else {
+            urlSchemeTask.didFailWithError(HandlerError.unsupportedScheme(url.scheme))
+            return
+        }
+
+        let href = url.absoluteString.removingPrefix(scheme + "://")
+        let resource = publication.get(href)
+        let task = Task(url: url, task: urlSchemeTask, resource: resource)
+        tasks[ObjectIdentifier(urlSchemeTask)] = task
```
Comment (ettore) [inline r2-navigator-swift/Toolkit/WebViewResourceHandler.swift:42]: access to this `tasks` dictionary doesn't seem to be thread safe: are these "start"/"stop" callbacks guaranteed to be called on the same thread? I'm unfamiliar with the `WKURLSchemeHandler` protocol but the docs don't seem to make any threading guarantees.

File: r2-navigator-swift/Toolkit/WebViewResourceHandler.swift [hunk 3]
```diff
@@ -91,39 +93,36 @@ final class WebViewResourceHandler: NSObject, WKURLSchemeHandler, Loggable {
...
 
                 self.resource.close()
             }
         }
 
         func cancel() {
-            ifNotCancelled {
-                isCancelled = true
-            }
+            assert(Thread.isMainThread)
```
[the remark this answers is not in the document]
Comment (mickael-menu) [reply r2-navigator-swift/Toolkit/WebViewResourceHandler.swift:111]: If I could fix `WKWebView`'s behavior, I would either:

* Not crash when calling APIs on a cancelled task, just ignore it.
* Take a `DispatchQueue` as argument that will be used to start/cancel tasks.

File: r2-navigator-swift/Toolkit/WebViewResourceHandler.swift [hunk 4]
```diff
@@ -91,39 +93,36 @@ final class WebViewResourceHandler: NSObject, WKURLSchemeHandler, Loggable {
...
+            self.isCancelled = true
         }
-
-        func ifNotCancelled(callback: () -> Void) {
-            isCancelledQueue.sync {
-                if !isCancelled {
-                    callback()
+        
+        /// Any API call to WKURLSchemeTask will crash the app if the task is cancelled by a call to:
```
[the remark this answers is not in the document]
Comment (mickael-menu) [reply r2-navigator-swift/Toolkit/WebViewResourceHandler.swift:115]: I think my other comment answers this too.

File: r2-navigator-swift/Toolkit/WebViewResourceHandler.swift (hunk 2 above)
Comment (mickael-menu) [reply r2-navigator-swift/Toolkit/WebViewResourceHandler.swift:42]: They seem to always be called from the main thread, but you are right that we can't make such assumption from the docs. I'll secure the access.

Thanks for taking the time to review this PR, this is a hairy one!

Comment (ettore) [reply r2-navigator-swift/Toolkit/WebViewResourceHandler.swift:42]: on second look... perhaps you should ignore my comment. These 2 start/stop callbacks are for a UIView subclass, so the expectation is that they are always called on the main thread. Therefore no multithreading issues exist on accessing the `tasks` dict from here.

File: r2-navigator-swift/Toolkit/WebViewResourceHandler.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ettore) [reply r2-navigator-swift/Toolkit/WebViewResourceHandler.swift:111]: I see, now I understand. Would you mind adding a link to this discussion in the code so that the reasoning behind these choices is not lost once you merge the code?

regarding the serial queue solution, would it be possible to schedule work on the serial queue `async`hronously instead of using `sync`?

[the remark this answers is not in the document]
Comment (mickael-menu) [reply r2-navigator-swift/Toolkit/WebViewResourceHandler.swift:111]: > I see, now I understand. Would you mind adding a link to this discussion in the code so that the reasoning behind these choices is not lost once you merge the code?

Sure, good idea.

> regarding the serial queue solution, would it be possible to schedule work on the serial queue `async`hronously instead of using `sync`?

The calls to `didReceive(Data)` must happen in sequence, otherwise the resources will be mangled in the web view. However, we probably could spawn multiple background tasks in sequence instead of a single background task with a for loop reading the bytes. That would probably solve the deadlock introduced with the custom serial queue.

But taking [your other comment](https://github.com/readium/r2-navigator-swift/pull/160#discussion_r648687252) into consideration, I wonder if using the main thread as the synchronization queue is not the proper approach here. I considered `WKURLSchemeHandler` to be background task, but it is true that it is coming from a UI framework. That explains why its APIs are coming from or going to (synchronously) the main thread.

[the remark this answers is not in the document]
Comment (ettore) [reply r2-navigator-swift/Toolkit/WebViewResourceHandler.swift:111]: > The calls to didReceive(Data) must happen in sequence, otherwise the resources will be mangled in the web view. However, we probably could spawn multiple background tasks in sequence instead of a single background task with a for loop reading the bytes. That would probably solve the deadlock introduced with the custom serial queue.

using the main thread to synchronize all the pieces is clever and _it does work_, so that's a big +1 and we should probably leave it in the commit history at the very least.

That being said we are doing various work on the main thread that ideally we'd do in a background thread. I don't have a clear idea about how expensive that work can be: if it _can't_ be very expensive, probably we could leave it on the main thread as-is and keep it simple. 

If it _can_ be expensive, and since there are pieces of work that have to be synchronous and sequential, if we can rewrite this code in a way that we can post the chunks of work we need _asynchronously_ to the serial queue, this would still preserve the sequential order and avoid deadlocks. One other equivalent approach could be to use NSOperations and chain dependencies explicitly but that's probably overkill. 

To summarize: using `sync` related to the main thread makes me nervous 😬. 

My understanding of `WKURLSchemeHandler` (very small, I have zero experience with it) is that yes it's part of a UI framework, but I wouldn't draw the conclusion that everything implicitly and reliably syncs on the main thread. I could be wrong though: the docs of that protocol are vague! We could ask a quick question at WWDC maybe in the [lounges](https://developer.apple.com/wwdc21/digital-lounges/) about whether we can expect its methods to _reliably_ be called on the main thread?

[the remark this answers is not in the document]
Comment (mickael-menu) [reply r2-navigator-swift/Toolkit/WebViewResourceHandler.swift:111]: > That being said we are doing various work on the main thread that ideally we'd do in a background thread. I don't have a clear idea about how expensive that work can be: if it can't be very expensive, probably we could leave it on the main thread as-is and keep it simple.

I think we're fine, there's no work done on the main thread except when calling back to `WKURLSchemeHandler` APIs (which is anyway calling to the main thread from my observations).

Using the main thread as the synchronization queue means that we never have to explicitly wait for something from the main thread.

If you look at the two delegate APIs (which are called from the main thread), they are not waiting on any queue:

https://github.com/readium/r2-navigator-swift/blob/6c603de421a7b5aa1c9d4aa8bc72be4c085fef0e/r2-navigator-swift/Toolkit/WebViewResourceHandler.swift#L29-L53


An improvement I could see is that instead of asserting we are (or not) on the main thread, we could schedule a main thread work when needed. This way if the `WKURLSchemeHandler` changes the used queue, it should still work. 

https://github.com/readium/r2-navigator-swift/blob/6c603de421a7b5aa1c9d4aa8bc72be4c085fef0e/r2-navigator-swift/Toolkit/WebViewResourceHandler.swift#L110-L126

[the remark this answers is not in the document]
Comment (ettore) [reply r2-navigator-swift/Toolkit/WebViewResourceHandler.swift:111]: the only example I believe is the one in `withTask(callback:)` and from your previous implementation with the custom queue `isCancelledQueue.sync { ... }`. If those could become `async`, that would be better since we'd at least remove the wait, and the order of WKURLSchemeTask api calls would still be preserved. But I think you said previously that that doesn't work for other reasons.

In any case, if we are sure that the work scheduled on the main thread is never going to be too much, I agree this should be fine.

[the remark this answers is not in the document]
Comment (mickael-menu) [reply r2-navigator-swift/Toolkit/WebViewResourceHandler.swift:111]: > the only example I believe is the one in `withTask(callback:)`

I think the closures are pretty harmless, but I can't guarantee that `WKURLSchemeHandler` will never do heavy work. Especially in `didReceive(Data)`. For now as we can see in the stacktrace, it is calling on the main thread synchronously anyway:



> and from your previous implementation with the custom queue `isCancelledQueue.sync { ... }`.

Yeah this was bad. If we ever go back to having a custom queue, we should do the approach we talked about earlier to replace the for loop.
